config/configauth/default_serverauthenticator.go

```diff
@@ -0,0 +1,127 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package configauth // import "go.opentelemetry.io/collector/config/configauth"
+
+import (
+	"context"
+	"net/http"
+
+	"google.golang.org/grpc"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/component/componenthelper"
+)
+
+var _ ServerAuthenticator = (*defaultServerAuthenticator)(nil)
+
+// Option represents the possible options for New.
+type Option func(*defaultServerAuthenticator)
+
+type defaultServerAuthenticator struct {
+	AuthenticateFunc
+	GRPCStreamInterceptorFunc
+	GRPCUnaryInterceptorFunc
+	HTTPInterceptorFunc
+
+	componenthelper.StartFunc
+	componenthelper.ShutdownFunc
+}
+
+// WithAuthenticate
+func WithAuthenticate(authenticateFunc AuthenticateFunc) Option {
+	return func(o *defaultServerAuthenticator) {
+		o.AuthenticateFunc = authenticateFunc
+	}
+}
+
+// WithGRPCStreamInterceptor
+func WithGRPCStreamInterceptor(grpcStreamInterceptorFunc GRPCStreamInterceptorFunc) Option {
+	return func(o *defaultServerAuthenticator) {
+		o.GRPCStreamInterceptorFunc = grpcStreamInterceptorFunc
+	}
+}
+
+// WithGRPCUnaryInterceptor
+func WithGRPCUnaryInterceptor(grpcUnaryInterceptorFunc GRPCUnaryInterceptorFunc) Option {
+	return func(o *defaultServerAuthenticator) {
+		o.GRPCUnaryInterceptorFunc = grpcUnaryInterceptorFunc
+	}
+}
+
+// WithHTTPInterceptor
+func WithHTTPInterceptor(httpInterceptorFunc HTTPInterceptorFunc) Option {
+	return func(o *defaultServerAuthenticator) {
+		o.HTTPInterceptorFunc = httpInterceptorFunc
+	}
+}
```

Comment:
@jpkrohling still not sure that the interceptors are useful. I feel that only the "AuthenticateFunc" is enough and everybody should implement only that. I would start with a smaller API then extend vs with a very generic API.

Reply:
ServerInterceptor is an API to both implementations (SPI) and consumers, in the form of configgrpc and confighttp. Unless each one of those would have the interceptors located on their side, and call the authenticate func on their own. Is that what you have in mind?

Reply:
Yes, more or less I see the interface becoming much simpler for the moment:
```golang
// ServerAuthenticator is an Extension that can be used as an authenticator for the configauth.Authentication option.
// Authenticators are then included as part of OpenTelemetry Collector builds and can be referenced by their
// names from the Authentication configuration. Each ServerAuthenticator is free to define its own behavior and configuration options,
// but note that the expectations that come as part of Extensions exist here as well. For instance, multiple instances of the same
// authenticator should be possible to exist under different names.
type ServerAuthenticator interface {
	component.Extension

	// Authenticate checks whether the given headers map contains valid auth data. Successfully authenticated calls will always return a nil error.
	// When the authentication fails, an error must be returned and the caller must not retry. This function is typically called from interceptors,
	// on behalf of receivers, but receivers can still call this directly if the usage of interceptors isn't suitable.
	// The deadline and cancellation given to this function must be respected, but note that authentication data has to be part of the map, not context.
	// The resulting context should contain the authentication data, such as the principal/username, group membership (if available), and the raw
	// authentication data (if possible). This will allow other components in the pipeline to make decisions based on that data, such as routing based
	// on tenancy as determined by the group membership, or passing through the authentication data to the next collector/backend.
	// The context keys to be used are not defined yet.
	Authenticate(ctx context.Context, headers map[string][]string) (context.Context, error)
}

// Then the helper funcs that you already have, things like:

// ToHTTPHandler provides a default implementation of http.Handler, useful for most authenticators.
// It passes the headers from the incoming request as it is, under the assumption that the credentials are part of it.
func ToHTTPHandler(authenticator ServerAuthenticator, next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		ctx, err := authenticator.Authenticate(r.Context(), r.Header)
		if err != nil {
			http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
			return
		}

		next.ServeHTTP(w, r.WithContext(ctx))
	})
}

```